hip_data_tools/apache/cassandra.py

```diff
@@ -0,0 +1,163 @@
+        batch = BatchStatement(consistency_level=ConsistencyLevel.QUORUM)
+        for row in rows:
+            batch.add(prepared_statement, row)
+        return batch
+
+    def _cql_upsert_from_dict(self, data, table):
+        upsert_sql = f"""
+        INSERT INTO {self.keyspace}.{table} 
+        ({", ".join(data[0])}) 
+        VALUES ({", ".join(['?' for key in data[0]])});
+            """
+        print(upsert_sql)
+        return upsert_sql
+
+    def _cql_upsert_from_dataframe(self, dataframe, table):
+        upsert_sql = f"""
+        INSERT INTO {self.keyspace}.{table} 
+        ({", ".join(list(dataframe.columns.values))}) 
+        VALUES ({", ".join(['?' for key in dataframe.columns.values])});
+            """
+        print(upsert_sql)
+        return upsert_sql
+
+    def upsert_dataframe(self, dataframe: DataFrame, table: str) -> None:
+        """
+        upload all data from a DataFrame onto a cassandra table
+        Args:
+            dataframe (DataFrame): a DataFrame to upsert
+            table (str): the table to upsert data into
+            column_mapping (dict): a mapping between column names of the dataframe to cassandra
+            table. If None then the DataFrame column names that match cassandra table anme will be
+            upserted else ignored
+        Returns: None
+        """
+        prepared_statement = self._session.prepare(
+            self._cql_upsert_from_dataframe(dataframe, table))
+        batch = self._prepare_batch(prepared_statement,
+                                    self._extract_rows_from_dataframe(dataframe))
+        return self._session.execute(batch)
+
+    def create_table_from_dataframe(self, dataframe: DataFrame, table: str) -> None:
+        """
+        create a table based on a given pandas DataFrame 's schema
+        Args:
+            dataframe (DataFrame):
+            table (str):
+        Returns: None
+        """
+        pass
+
+    def upsert_dict(self, data: list, table: str) -> ResultSet:
+        """
+        Upsert a row into a given cassandra table based on the dictionary key values
+        Args:
+            data (list[dict]):
+            table (str):
+        Returns: None
+        """
+        prepared_statement = self._session.prepare(self._cql_upsert_from_dict(data, table))
+        batch = self._prepare_batch(prepared_statement, self._extract_rows_from_list_of_dict(data))
+        return self._session.execute(batch)
+
+    def create_table_from_model(self, model_class):
+        """
+        create a table if not exists from the given Model Class
+        Args:
+            model_class (class): class for the Model
+        Returns: None
+        """
+        self._conn.setup_connection(default_keyspace=self.keyspace)
+        sync_table(model_class)
+
+    def create_table(self, columns: dict, primary_keys: list, table: str) -> ResultSet:
+        """
+        Create a table if not already exists based on data from a dictionary
```

Comment:
are you capitalising the start of comments or not? can't remember the standard for this repo?

Reply:
I am. We should.